webhooks/finished/server.js

```diff
@@ -0,0 +1,407 @@
+          );
+        }
+      } else if (product === "TRANSACTIONS") {
+        // Most of these aren't needed if you use the new sync API!
+        if (code === "INITIAL_UPDATE") {
+          console.log(
+            `First patch of transactions are done. There's ${req.body.new_transactions} available`
+          );
+        } else if (code === "HISTORICAL_UPDATE") {
+          console.log(
+            `Historical transactions are done. There's ${req.body.new_transactions} available`
+          );
+        } else if (code === "DEFAULT_UPDATE") {
+          console.log(
+            `New data is here! There's ${req.body.new_transactions} available`
+          );
+        } else if (code === "TRANSACTIONS_REMOVED") {
+          console.log(
+            `Looks like a few transactions have been reversed and should be removed from our records`
+          );
+        }
+      }
+
+      res.json({ status: "received" });
+    } else {
+      console.log("Webhook didn't pass verification!");
+      res.status(401);
+    }
+  } catch (error) {
+    next(error);
+  }
+});
+
+const verifyWebhook = async (req) => {
+  const verbose = false;
+  try {
+    const signedJwt = req.headers["plaid-verification"];
+    const decodedToken = jwt_decode(signedJwt);
+    // Extract the JWT header
+    const decodedTokenHeader = jwt_decode(signedJwt, { header: true });
+    // Extract the kid value from the header
+    verbose &&
+      console.log(
+        `Your token is ${JSON.stringify(
+          decodedToken
+        )} with headers ${JSON.stringify(decodedTokenHeader)}`
+      );
+    const currentKeyID = decodedTokenHeader.kid;
+    verbose && console.log(`Key ID is ${currentKeyID}`);
+    const keyResponse = await plaidClient.webhookVerificationKeyGet({
+      key_id: currentKeyID,
+    });
+    verbose &&
+      console.log(`Your key from Plaid is ${JSON.stringify(keyResponse.data)}`);
+    // TODO: Cache this so that I don't need to call this every time we receive
+    // a webhook.
+    const key = keyResponse.data["key"];
+    const keyLike = await JWT.importJWK(key);
+    const { payload } = await JWT.jwtVerify(signedJwt, keyLike, {
+      maxTokenAge: "5 min",
+    });
+    verbose && console.log("JWT is verified");
+    // `payload` should be the same as the decoded token body.
+    verbose && console.log(payload);
+    const requestBody = JSON.stringify(req.body);
+    verbose && console.log(`Your body is ${requestBody}`);
+    const bodyHash = req.bodyHash;
+    const claimedBodyHash = payload.request_body_sha256;
+    verbose &&
+      console.log(
+        `Body has is ${bodyHash} compared to what I got in the header ${claimedBodyHash}`
+      );
+    return compare(bodyHash, claimedBodyHash);
+  } catch (error) {
+    console.log(`Received an error attempting to verify the webhook:`);
+    console.log(error);
+    return false;
+  }
+};
+
+webhookApp.use(errorHandler);
```

Comment:
nicely done.  Remind me to steal this code for the pattern apps to verify webhooks.

Reply:
Feel free! But for the Pattern apps, you probably do want to cache the keys. (Called out in line 381)